File: src/field/grid.ts
/**
 * The field grid.
 *
 * Tiles carry an `elevation` as well as walkability. That single extra number
 * is what lets a 2.5D renderer stack a map into layers - a bridge over a
 * river, a ledge you can drop off but not climb - while the rules layer stays
 * a flat 2D array. Movement between tiles is allowed when the height
 * difference is within `stepHeight`; dropping further is allowed but climbing
 * is not, which is the classic ledge rule and costs nothing to express.
 */

export interface Tile {
  /** Blocks movement outright regardless of height. */
  solid: boolean;
  /** Height in tile units. */
  elevation: number;
  /** Free-form terrain tag the renderer and abilities can key off. */
  terrain?: string;
  /** Blocks line of sight for targeting and vision cones. */
  opaque?: boolean;
}

export interface Vec2 {
  x: number;
  y: number;
}

export type Facing = 'north' | 'south' | 'east' | 'west';

export const FACING_VECTORS: Readonly<Record<Facing, Vec2>> = Object.freeze({
  north: { x: 0, y: -1 },
  south: { x: 0, y: 1 },
  east: { x: 1, y: 0 },
  west: { x: -1, y: 0 },
});

export interface GridOptions {
  /** Maximum climbable height difference between adjacent tiles. */
  stepHeight?: number;
  /** Maximum drop before movement is blocked. Infinity allows any fall. */
  maxDrop?: number;
}

export class Grid {
  readonly width: number;
  readonly height: number;
  readonly stepHeight: number;
  readonly maxDrop: number;
  private readonly tiles: Tile[];

  constructor(width: number, height: number, tiles?: Tile[], options: GridOptions = {}) {
    this.width = width;
    this.height = height;
    this.stepHeight = options.stepHeight ?? 1;
    this.maxDrop = options.maxDrop ?? Number.POSITIVE_INFINITY;
    this.tiles =
      tiles ?? Array.from({ length: width * height }, () => ({ solid: false, elevation: 0 }) satisfies Tile);
    if (this.tiles.length !== width * height) {
      throw new Error(`Grid expects ${width * height} tiles, received ${this.tiles.length}`);
    }
  }

  inBounds(x: number, y: number): boolean {
    return x >= 0 && y >= 0 && x < this.width && y < this.height;
  }

  at(x: number, y: number): Tile | undefined {
    if (!this.inBounds(x, y)) return undefined;
    return this.tiles[y * this.width + x];
  }

  set(x: number, y: number, tile: Tile): void {
    if (!this.inBounds(x, y)) return;
    this.tiles[y * this.width + x] = tile;
  }

  /** Mutate one field of a tile, for abilities that raise or lower terrain. */
  patch(x: number, y: number, patch: Partial<Tile>): void {
    const tile = this.at(x, y);
    if (!tile) return;
    this.set(x, y, { ...tile, ...patch });
  }

  /**
   * Whether an actor standing on `from` may step onto `to`. Climbing is capped
   * by `stepHeight`; falling is capped by `maxDrop`.
   */
  canStep(from: Vec2, to: Vec2): boolean {
    const source = this.at(from.x, from.y);
    const target = this.at(to.x, to.y);
    if (!source || !target || target.solid) return false;
    const delta = target.elevation - source.elevation;
    if (delta > this.stepHeight) return false;
    if (-delta > this.maxDrop) return false;
    return true;
  }

  /** Straight-line visibility, used for targeting and NPC sight cones. */
  hasLineOfSight(from: Vec2, to: Vec2): boolean {
    for (const point of bresenham(from, to)) {
      if (point.x === from.x && point.y === from.y) continue;
      if (point.x === to.x && point.y === to.y) continue;
      const tile = this.at(point.x, point.y);
      if (!tile || tile.solid || tile.opaque) return false;
    }
    return true;
  }

  /** Serializable form, for save files and for shipping maps as JSON. */
  toJSON(): { width: number; height: number; stepHeight: number; maxDrop: number; tiles: Tile[] } {
    return {
      width: this.width,
      height: this.height,
      stepHeight: this.stepHeight,
      maxDrop: this.maxDrop === Number.POSITIVE_INFINITY ? -1 : this.maxDrop,
      tiles: this.tiles.map((tile) => ({ ...tile })),
    };
  }

  static fromJSON(data: ReturnType<Grid['toJSON']>): Grid {
    return new Grid(data.width, data.height, data.tiles, {
      stepHeight: data.stepHeight,
      maxDrop: data.maxDrop < 0 ? Number.POSITIVE_INFINITY : data.maxDrop,
    });
  }
}

export function translate(position: Vec2, facing: Facing, distance = 1): Vec2 {
  const vector = FACING_VECTORS[facing];
  return { x: position.x + vector.x * distance, y: position.y + vector.y * distance };
}

export function manhattan(a: Vec2, b: Vec2): number {
  return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
}

export function samePosition(a: Vec2, b: Vec2): boolean {
  return a.x === b.x && a.y === b.y;
}

/** Integer line walk, inclusive of both endpoints. */
export function bresenham(from: Vec2, to: Vec2): Vec2[] {
  const points: Vec2[] = [];
  let x0 = from.x;
  let y0 = from.y;
  const dx = Math.abs(to.x - x0);
  const dy = -Math.abs(to.y - y0);
  const sx = x0 < to.x ? 1 : -1;
  const sy = y0 < to.y ? 1 : -1;
  let error = dx + dy;

  for (;;) {
    points.push({ x: x0, y: y0 });
    if (x0 === to.x && y0 === to.y) break;
    const doubled = 2 * error;
    if (doubled >= dy) {
      error += dy;
      x0 += sx;
    }
    if (doubled <= dx) {
      error += dx;
      y0 += sy;
    }
  }
  return points;
}

File: unity/Aetherlight.Core/Field/Grid.cs
using System;
using System.Collections.Generic;

namespace Aetherlight.Field
{
    public struct Vec2 : IEquatable<Vec2>
    {
        public int X;
        public int Y;

        public Vec2(int x, int y)
        {
            X = x;
            Y = y;
        }

        public bool Equals(Vec2 other) => X == other.X && Y == other.Y;
        public override bool Equals(object? obj) => obj is Vec2 other && Equals(other);
        public override int GetHashCode() => unchecked(X * 397 ^ Y);
        public override string ToString() => $"({X}, {Y})";

        public static bool operator ==(Vec2 a, Vec2 b) => a.Equals(b);
        public static bool operator !=(Vec2 a, Vec2 b) => !a.Equals(b);
    }

    public enum Facing { North, South, East, West }

    public sealed class Tile
    {
        /// <summary>Blocks movement outright regardless of height.</summary>
        public bool Solid;

        /// <summary>Height in tile units.</summary>
        public double Elevation;

        /// <summary>Free-form terrain tag the renderer and abilities key off.</summary>
        public string? Terrain;

        /// <summary>Blocks line of sight for targeting and vision cones.</summary>
        public bool Opaque;

        public Tile Clone() => new Tile { Solid = Solid, Elevation = Elevation, Terrain = Terrain, Opaque = Opaque };
    }

    /// <summary>
    /// The field grid.
    ///
    /// Tiles carry an elevation as well as walkability. That single extra
    /// number is what lets a 2.5D renderer stack a map into layers - a bridge
    /// over a river, a ledge you can drop off but not climb - while the rules
    /// layer stays a flat 2D array. Movement between tiles is allowed when the
    /// height difference is within StepHeight; dropping further is allowed but
    /// climbing is not, which is the classic ledge rule and costs nothing to
    /// express.
    /// </summary>
    public sealed class Grid
    {
        public int Width { get; }
        public int Height { get; }

        /// <summary>Maximum climbable height difference between adjacent tiles.</summary>
        public double StepHeight { get; }

        /// <summary>Maximum drop before movement is blocked. Infinity allows any fall.</summary>
        public double MaxDrop { get; }

        private readonly Tile[] _tiles;

        public Grid(int width, int height, IReadOnlyList<Tile>? tiles = null, double stepHeight = 1, double maxDrop = double.PositiveInfinity)
        {
            Width = width;
            Height = height;
            StepHeight = stepHeight;
            MaxDrop = maxDrop;

            if (tiles == null)
            {
                _tiles = new Tile[width * height];
                for (int i = 0; i < _tiles.Length; i++) _tiles[i] = new Tile();
            }
            else
            {
                if (tiles.Count != width * height)
                    throw new ArgumentException($"Grid expects {width * height} tiles, received {tiles.Count}", nameof(tiles));
                _tiles = new Tile[tiles.Count];
                for (int i = 0; i < tiles.Count; i++) _tiles[i] = tiles[i];
            }
        }

        public bool InBounds(int x, int y) => x >= 0 && y >= 0 && x < Width && y < Height;

        public Tile? At(int x, int y) => InBounds(x, y) ? _tiles[y * Width + x] : null;

        public Tile? At(Vec2 position) => At(position.X, position.Y);

        public void Set(int x, int y, Tile tile)
        {
            if (!InBounds(x, y)) return;
            _tiles[y * Width + x] = tile;
        }

        /// <summary>Mutate part of a tile, for abilities that reshape terrain.</summary>
        public void Patch(int x, int y, bool? solid = null, double? elevation = null, string? terrain = null, bool? opaque = null)
        {
            var tile = At(x, y);
            if (tile == null) return;
            if (solid.HasValue) tile.Solid = solid.Value;
            if (elevation.HasValue) tile.Elevation = elevation.Value;
            if (terrain != null) tile.Terrain = terrain;
            if (opaque.HasValue) tile.Opaque = opaque.Value;
        }

        /// <summary>
        /// Whether an actor standing on <paramref name="from"/> may step onto
        /// <paramref name="to"/>. Climbing is capped by StepHeight; falling is
        /// capped by MaxDrop.
        /// </summary>
        public bool CanStep(Vec2 from, Vec2 to)
        {
            var source = At(from);
            var target = At(to);
            if (source == null || target == null || target.Solid) return false;

            double delta = target.Elevation - source.Elevation;
            if (delta > StepHeight) return false;
            if (-delta > MaxDrop) return false;
            return true;
        }

        /// <summary>Straight-line visibility, for targeting and NPC sight cones.</summary>
        public bool HasLineOfSight(Vec2 from, Vec2 to)
        {
            foreach (var point in Bresenham(from, to))
            {
                if (point == from || point == to) continue;
                var tile = At(point);
                if (tile == null || tile.Solid || tile.Opaque) return false;
            }
            return true;
        }

        /// <summary>Serializable form, for save files and shipping maps as JSON.</summary>
        public GridData ToData()
        {
            var data = new GridData
            {
                Width = Width,
                Height = Height,
                StepHeight = StepHeight,
                // JSON has no infinity, so the sentinel travels as -1.
                MaxDrop = double.IsPositiveInfinity(MaxDrop) ? -1 : MaxDrop,
            };
            foreach (var tile in _tiles) data.Tiles.Add(tile.Clone());
            return data;
        }

        public static Grid FromData(GridData data) => new Grid(
            data.Width,
            data.Height,
            data.Tiles,
            data.StepHeight,
            data.MaxDrop < 0 ? double.PositiveInfinity : data.MaxDrop);

        // --- free functions --------------------------------------------------

        public static readonly IReadOnlyDictionary<Facing, Vec2> FacingVectors = new Dictionary<Facing, Vec2>
        {
            [Facing.North] = new Vec2(0, -1),
            [Facing.South] = new Vec2(0, 1),
            [Facing.East] = new Vec2(1, 0),
            [Facing.West] = new Vec2(-1, 0),
        };

        public static Vec2 Translate(Vec2 position, Facing facing, int distance = 1)
        {
            var vector = FacingVectors[facing];
            return new Vec2(position.X + vector.X * distance, position.Y + vector.Y * distance);
        }

        public static int Manhattan(Vec2 a, Vec2 b) => Math.Abs(a.X - b.X) + Math.Abs(a.Y - b.Y);

        /// <summary>Integer line walk, inclusive of both endpoints.</summary>
        public static List<Vec2> Bresenham(Vec2 from, Vec2 to)
        {
            var points = new List<Vec2>();
            int x0 = from.X;
            int y0 = from.Y;
            int dx = Math.Abs(to.X - x0);
            int dy = -Math.Abs(to.Y - y0);
            int sx = x0 < to.X ? 1 : -1;
            int sy = y0 < to.Y ? 1 : -1;
            int error = dx + dy;

            while (true)
            {
                points.Add(new Vec2(x0, y0));
                if (x0 == to.X && y0 == to.Y) break;
                int doubled = 2 * error;
                if (doubled >= dy)
                {
                    error += dy;
                    x0 += sx;
                }
                if (doubled <= dx)
                {
                    error += dx;
                    y0 += sy;
                }
            }
            return points;
        }
    }

    public sealed class GridData
    {
        public int Width;
        public int Height;
        public double StepHeight = 1;
        /// <summary>-1 stands in for "no limit", since JSON has no infinity.</summary>
        public double MaxDrop = -1;
        public List<Tile> Tiles = new List<Tile>();
    }
}
